delta-lake/delta-spark341db/pom.xml

```diff
@@ -258,11 +62,10 @@
             <version>${spark.version}</version>
             <scope>provided</scope>
         </dependency>
-            <dependency>
+        <dependency>
             <groupId>org.apache.spark</groupId>
             <artifactId>spark-sql-api_${scala.binary.version}</artifactId>
             <version>${spark.version}</version>
-            <scope>provided</scope>
```

Comment:
I shouldn't have removed "provided", that was a mistake.

Regarding the other question, When I remove this block in this file, the build fails with:

```
[ERROR] [Error] /home/ubuntu/spark-rapids/delta-lake/common/src/main/databricks/scala/com/databricks/sql/transaction/tahoe/rapids/GpuDeltaCatalogBase.scala:121: Symbol 'type org.apache.spark.sql.SqlApiConf' is missing from the classpath.
This symbol is required by 'class org.apache.spark.sql.internal.SQLConf'.
```

My understanding is that the transitive dependency would not be sufficient in the case that both legs are of provided scope (so `rapids-4-spark-delta-spark341db_2.12` compilation does not include `spark-sql-api` in the classpath). I am also assuming that in the prior spark versions, SQLConf does not depend on SqlApiConf (that's just one example of several similar errors).

Reply:
So, the SqlApiConf is a new class that didn't exist in prior versions. So it is a new dependency but Jason brings up a good point that the dependency should've been transitively included, I don't know enough about the way maven handles dependencies with provided scope to answer this.

Reply:
The maven docs seem to agree that you can't have transitive dependencies of provided scope, or at least that such dependencies won't be propagated as you might expect

https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope

Reply:
My main point still applies.  Accessing org.apache.spark.sql.internal.SQLConf is not something unique to the delta-spark341db project, so why does it need this additional dependency but other projects do not?  The whole point of this PR is to make the Databricks dependency stuff consistent, and I'm wondering how pulling in the rapids-4-spark-db-bom dependency isn't bringing the needed spark-sql-api dependency.  Is it a problem where rapids-4-spark-db-bom is supposed to be asking for this on spark341db but is not?  Or put another way, why isn't this dependency appearing in a ton of other poms yet it compiles there?

I think the issue is that the delta shim poms shiould not be using the jdk profiles project as the parent but instead the shim-deps project as the parent.  That's what most other projects are doing, and I suspect would fix this odd, extra dependency here.

Reply:
The spark-sql-api jar literally doesn't exist on earlier spark versions. In databricks, we fetch these dependencies from the local compute node. On spark 3.4.1 when we build we can see
```
Generating an execution for org.apache.spark spark-sql-api_2.12 ----ws_3_4--sql--api--sql-api-hive-2.3__hadoop-3.2_2.12_deploy.jar
```
which I believe is copying the needed jar into our local `~/.m2`

On a 330 runtime node we can see that jar is not there:
```
ubuntu@0815-154041-g4nf7g3z-10-59-189-224:~/spark-rapids$ ls /databricks/jars/----ws_3_3--sql--
----ws_3_3--sql--catalyst--catalyst-hive-2.3__hadoop-3.2_2.12_deploy.jar
----ws_3_3--sql--catalyst--libcatalyst_resources.jar
----ws_3_3--sql--catalyst--libspark-sql-parser-compiled.jar
----ws_3_3--sql--catalyst--proto-hive-2.3__hadoop-3.2_2.12_deploy.jar
----ws_3_3--sql--core--core-hive-2.3__hadoop-3.2_2.12_deploy.jar
----ws_3_3--sql--core--libcore_resources.jar
----ws_3_3--sql--core--proto_2.12_deploy.jar
----ws_3_3--sql--hive--hive-hive-2.3__hadoop-3.2_2.12_deploy_shaded.jar
----ws_3_3--sql--hive--libhive_resources.jar
----ws_3_3--sql--hive--org.apache.commons__commons-pool2__2.11.1_shaded-for-hive-hive-2.3__hadoop-3.2.jar
----ws_3_3--sql--hive-thriftserver--hive-thriftserver-hive-2.3__hadoop-3.2_2.12_deploy.jar
----ws_3_3--sql--hive-thriftserver--hive-thriftserver-protocol-hive-2.3__hadoop-3.2_2.12_deploy.jar
```

Therefore if we tried to include this as a shared dependency we would get:
```
[ERROR] Failed to execute goal on project rapids-4-spark-db-bom: Could not resolve dependencies for project com.nvidia:rapids-4-spark-db-bom:pom:24.10.0-SNAPSHOT: Could not find artifact org.apache.spark:spark-sql-api_2.12:jar:3.3.0-databricks
```

So even though I agree the point is to (mostly) consolidate redundant dependencies, I don't see how it could apply here. 

> Accessing org.apache.spark.sql.internal.SQLConf is not something unique to the delta-spark341db project
That's true but that class is not in the spark-sql-api, and the difference is that that class itself started to depend on `org.apache.spark.sql.SqlApiConf`

Reply:
> The spark-sql-api jar literally doesn't exist on earlier spark versions

Yes, I realize that.  But it's interesting that many projects that need to build on Databricks don't explicitly include this dependency, yet this one does.  The main difference I see is those projects use shim-deps as their parent, but the delta shims do not.  Makes me wonder if that's the correct fix here.

Reply:
Ah you are right, that does fix it. I pushed a commit to make that change. I guess it makes sense because now the scope is provided, so it won't complain if it's not there, and also since it's retrieved via the parent inheritance rather than as a transitive dependency, it will have the correct classpath.

Reply:
Cool, that looks better.  Curious, is the rapids-4-spark-db-bom dependency still needed?  Only the delta databricks shims are explicitly including it (besidses shim-deps, of course), and I suspect it's no longer necessary after switching to the proper parent pom.

Reply:
> also since it's retrieved via the parent inheritance rather than as a transitive dependency, it will have the correct classpath

To clarify here, transitive dependencies do indeed work.  We rely on them all the time in our build.  The only reason they don't work for Databricks builds is because we're manually adding Databricks jars to the project without a pom, and thus Maven doesn't see any transitive dependencies to pick up.  That's why the rapids-4-spark-db pom lists all sorts of dependencies that aren't seen anywhere else in the project.  It's not that those dependencies aren't used or needed outside of a Databricks build.  It's because only in a Databricks build we're not getting these dependencies automatically and thus must request them manually.

Reply:
> Curious, is the rapids-4-spark-db-bom dependency still needed?

I think not, just tested with one build. I removed those as well